### DISCUSSION TITLE
How to improve performance?

### DISCUSSION THREAD

**Discussion by ShaonDey** [MarkPflug/Sylvan#227] [Q&A] [answered]
Using:
```
<TargetFramework>net6.0</TargetFramework>
<PackageReference Include="FluentValidation.AspNetCore" Version="11.2.2" />
<PackageReference Include="Sylvan.Data.Csv" Version="1.3.5" />
```

</br>

The caller will upload recipient information in a .CSV file as:
```
Id,Reference,Amount
1,123456789,30
18,123-456-789,60.25
```

</br>

Recipient model:
```
public class Recipient
{
    public long ProductId { get; set; }
    public string AccountReference { get; set; }
    public decimal Amount { get; set; }
}
```

</br>

API:
```
public async Task UploadRecipients([FromForm] IFormFile csvFormFile, CancellationToken cancellationToken)
{
    var mappedRecipients = new List<Recipient>();
    using var streamReader = new StreamReader(csvFormFile.OpenReadStream());
    var options = new CsvDataReaderOptions
    {
        HasHeaders = true,
        Delimiter = ',',
        CsvStyle = CsvStyle.Escaped,
        MaxBufferSize = int.MaxValue,
        Culture = CultureInfo.InvariantCulture
    };
    await using var reader = await CsvDataReader.CreateAsync(streamReader, options, cancellationToken);
    while (await reader.ReadAsync(cancellationToken))
    {
        var item = new Recipient
        {
            ProductId = reader.GetInt64(0),
            AccountReference = reader.GetString(1),
            Amount = reader.GetDecimal(2)
        };
        await _fluentValidator.ValidateAndThrowAsync(item, cancellationToken);
        mappedRecipients.Add(item);
    }
}
```

</br></br>

Consider `await _fluentValidator.ValidateAndThrowAsync(item, cancellationToken);` line be absolutely necessary for validation purposes. Currently, it's taking 60 milliseconds on average to process a CSV file of 1.76 MB in size with 100001 rows (including one header on top). Now can this piece of code blocks be improved to be more performant?

**Comment by MarkPflug** [comment]
You can initialize the mappedRecipients list to have a larger initial capacity if you know the number of records will be large. This would avoid having to resize the internal buffer repeatedly, but probably won't have a significant impact on performance. Otherwise, this looks pretty optimal. How fast is it with the fluentValidator bit removed?

**Reply by ShaonDey** [reply, marked as answer]
Initializing the mappedRecipients list with an initial capacity beforehand is not possible as I do not know the number of records. You can consider the fluentValidator is performing at its best. I'd probably choose between the options I have.
